pkg/cloud-provider/cloudapi/azure/azure_security.go

```diff
@@ -83,51 +113,58 @@ func (computeCfg *computeServiceConfig) processAppliedToMembership(appliedToGrou
 	}
 
 	// find network interfaces which are using or need to use the provided NSG
-	nwIntfIDSetNsgToAttach := make(map[string]struct{})
-	nwIntfIDSetNsgToDetach := make(map[string]struct{})
+	nwIntfIDSetNsgToAttach := make(map[string]*armnetwork.Interface)
+	nwIntfIDSetNsgToDetach := make(map[string]*armnetwork.Interface)
 	for _, networkInterface := range networkInterfaces {
 		nwIntfIDLowerCase := strings.ToLower(*networkInterface.ID)
 		// 	for network interfaces not attached to any virtual machines, skip processing
-		if emptyString(networkInterface.VirtualMachineID) {
+		if networkInterface.Properties.VirtualMachine == nil || emptyString(networkInterface.Properties.VirtualMachine.ID) {
 			continue
 		}
 
 		isNsgAttached := false
-		if networkInterface.NetworkSecurityGroupID != nil && len(*networkInterface.NetworkSecurityGroupID) > 0 {
-			nsgID := strings.ToLower(*networkInterface.NetworkSecurityGroupID)
+		if networkInterface.Properties.NetworkSecurityGroup != nil && !emptyString(networkInterface.Properties.NetworkSecurityGroup.ID) {
+			nsgID := strings.ToLower(*networkInterface.Properties.NetworkSecurityGroup.ID)
 			_, _, nsgNameLowercase, err := extractFieldsFromAzureResourceID(nsgID)
 			if err != nil {
 				azurePluginLogger().Error(err, "nsg ID format not valid", "nsgID", nsgID)
 				return err
 			}
 
-			attachedAsgIDs := networkInterface.ApplicationSecurityGroupIDs
-			for _, asgID := range attachedAsgIDs {
-				asgIDLowerCase := strings.ToLower(*asgID)
-				// proceed only if network-interface attached to nephe ASG
-				_, _, asgName, err := extractFieldsFromAzureResourceID(asgIDLowerCase)
-				if err != nil {
+			for _, ipConfigs := range networkInterface.Properties.IPConfigurations {
+				if ipConfigs.Properties == nil {
 					continue
 				}
-				_, _, isAT := securitygroup.IsNepheControllerCreatedSG(asgName)
-				if isAT {
-					if strings.Compare(nsgNameLowercase, perVnetNsgNameLowercase) == 0 &&
-						strings.Compare(cloudSgNameLowercase, asgName) == 0 {
-						isNsgAttached = true
-						break
+				for _, asg := range ipConfigs.Properties.ApplicationSecurityGroups {
+					asgIDLowerCase := strings.ToLower(*asg.ID)
+					// proceed only if network-interface attached to nephe ASG
+					_, _, asgName, err := extractFieldsFromAzureResourceID(asgIDLowerCase)
+					if err != nil {
+						continue
+					}
+					_, _, isAT := securitygroup.IsNepheControllerCreatedSG(asgName)
+					if isAT {
+						if strings.Compare(nsgNameLowercase, perVnetNsgNameLowercase) == 0 &&
+							strings.Compare(cloudSgNameLowercase, asgName) == 0 {
+							isNsgAttached = true
+							break
+						}
 					}
 				}
+				// Handling just first available ip-configuration; with an assumption that it will be primary IP always.
```

Comment:
Is this a TODO? or currently does the first index always returns primary IP config?

Reply:
yes.. I verified